src/cdomains/apron/sharedFunctions.apron.ml

```diff
@@ -112,44 +146,164 @@ struct
+                | true -> texpr1_expr_of_cil_exp e
                 | false
                 | exception Cilfacade.TypeOfError _ (* typeOf inner e, not outer exp *)
                 | exception Invalid_argument _ -> (* get_ikind in is_cast_injective *)
                   raise (Unsupported_CilExp (Cast_not_injective t))
               end
             | _ ->
               raise (Unsupported_CilExp Exp_not_supported)
-          in
-          if not no_ov then (
-            let (type_min, type_max) = IntDomain.Size.range ik in
-            let texpr1 = Texpr1.of_expr env expr in
-            match Bounds.bound_texpr d texpr1 with
-            | Some min, Some max when Z.compare type_min min <= 0 && Z.compare max type_max <= 0 -> ()
-            | min_opt, max_opt ->
-              if M.tracing then M.trace "apron" "may overflow: %a (%a, %a)\n" CilType.Exp.pretty exp (Pretty.docOpt (IntOps.BigIntOps.pretty ())) min_opt (Pretty.docOpt (IntOps.BigIntOps.pretty ())) max_opt;
-              raise (Unsupported_CilExp Overflow)
-          );
+          in overflow_handling_apron no_ov ik env expr d exp;
           expr
         | exception (Cilfacade.TypeOfError _ as e)
         | exception (Invalid_argument _ as e) ->
           raise (Unsupported_CilExp (Exp_typeOf e))
     in
     texpr1_expr_of_cil_exp exp
 
-  let texpr1_of_cil_exp d env e no_ov =
+
+  let texpr1_expr_of_cil_exp_with_overflow_check (ask: Queries.ask) d env exp no_ov overflow_handling =
+    let query e ik =
+      let res = match ask.f (EvalInt e) with
+        | `Bot -> raise (Unsupported_CilExp Exp_not_supported) (* This should never happen according to Michael Schwarz *)
+        | `Top -> IntDomain.IntDomTuple.top_of ik
+        | `Lifted x -> x (* According to base.ml:704 cast should be unnecessary because it should be taken care of by EvalInt. *)
+      in
+      (* If the returned interval is top of the expected ikind (i.e. the value is unknown ) or the returned interval is in range of the expected interval, return top
+         - If top is returned the expression will be rewritten.
+         - If a constant is returned this specific value is casted to the expected ikind value
+         - else we got an interval with unsupported bounds i.e. the value expression is known to be unknown and needs casting, which we do not support i.e. the expression is not supported*)
+      let top = IntDomain.IntDomTuple.is_top_of ik res in
+      (* || (Option.is_some min && Option.is_some max && (minimal  != maximal && (Option.get min >= minimal || Option.get max <= maximal))) in*)
+      if top then IntDomain.IntDomTuple.top_of ik else res
+    in
+    (* recurse without env argument *)
+    let rec texpr1_expr_of_cil_exp ask exp =
+      match exp with
+      | Lval (Var v, NoOffset) when Tracked.varinfo_tracked v ->
+        if not v.vglob || Arg.allow_global then
```

Comment:
For advisors to check: Is `vglob` correct or would we have to use a query here?

Reply:
This has been _copied_ from above from our existing code, so it should be right. But I'm more concerned about the copying here.